### DISCUSSION TITLE
buffer object leaking memory

### DISCUSSION THREAD

**Discussion by janmarcheckman** [gfx-rs/wgpu#5918] [Q&A]
I have a program that is leaking memory.
It seems to be related to the fact that I'm creating a buffer every frame.
As far as I understand wgpu objects should free their resources on drop.
Am I correct that this shouldn't leak memory? Or am I missing something?

I've written the simplest program I could that still exhibits the behavior:

```rust
fn main() {
    use wgpu::util::DeviceExt;

    let instance = wgpu::Instance::new(wgpu::InstanceDescriptor {
        backends: wgpu::Backends::all(),
        ..Default::default()
    });
    
    let adapter = pollster::block_on(instance.request_adapter(&wgpu::RequestAdapterOptions {
        power_preference: wgpu::PowerPreference::HighPerformance,
        force_fallback_adapter: false,
        compatible_surface: None,
    })).unwrap();

    let (device, _queue) = pollster::block_on(adapter.request_device(&wgpu::DeviceDescriptor {
        label: None,
        required_features: wgpu::Features::empty(),
        required_limits: wgpu::Limits::default()
    }, None)).unwrap();

    let contents = &[0, 0, 0, 0];

    let buffer_init_descriptor = &wgpu::util::BufferInitDescriptor {
        label: None,
        contents,
        usage: wgpu::BufferUsages::UNIFORM,
    };

    loop {
        let _buffer = device.create_buffer_init(buffer_init_descriptor);
    }
}
```

```toml
[package]
name = "test_buffer_leak"
version = "0.1.0"
edition = "2021"

[dependencies]
wgpu = "0.20.1"
pollster = "0.3.0"
```

I'm on windows 10
The backend that gets selected is Vulkan

**Comment by teoxoy** [comment]
The destruction is currently deferred to `Queue.submit()`/`Surface.configure()`/`Device.poll()`, with https://github.com/gfx-rs/wgpu/pull/5910 it will happen either immediately if there are no submissions using that resource or otherwise deferred at the end of the submission.

**Reply by janmarcheckman** [reply]
Thank you for your reply! :)

I've modified the loop to be:

```rust
loop {
    let _buffer = device.create_buffer_init(buffer_init_descriptor);
    device.poll(wgpu::MaintainBase::Wait);
}
```

And the program is still leaking memory.
Am I still missing something?

**Reply by teoxoy** [reply]
Maybe there was a bug preventing destruction; now that https://github.com/gfx-rs/wgpu/pull/5910 has been merged, could you try to reproduce the leak on current trunk?

**Reply by teoxoy** [reply]
With https://github.com/gfx-rs/wgpu/pull/5910 the poll should no longer be necessary either in your example.

**Reply by janmarcheckman** [reply]
I've updated the dependency to:
```toml
wgpu = { git = "https://github.com/gfx-rs/wgpu" }
```
But the leak is still there

**Reply by teoxoy** [reply]
We ended up closing https://github.com/gfx-rs/wgpu/issues/6053 as "won't fix". Just to confirm: are you still seeing the leak if you call `Queue.submit()`?

**Reply by janmarcheckman** [reply]
I've changed the loop to be:

``` rust
loop {
    let _buffer = device.create_buffer_init(buffer_init_descriptor);
    let encoder = device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
    queue.submit(Some(encoder.finish()));
}
```

And I'm still seeing the leak

**Reply by teoxoy** [reply]
I'm seeing the leak on v0.20 but not on v22. Are you on v22?

**Reply by janmarcheckman** [reply]
My cargo lock file specifies:

[[package]]
name = "wgpu"
version = "22.0.0"
source = "git+https://github.com/gfx-rs/wgpu#9c6ae1beae2490ce44d99034f7f1faada936f3d8"

And I'm still seeing the leak.

**Reply by teoxoy** [reply]
Ok, you previously mentioned win 10, vulkan, is it a release or debug build?

How big is the leak you are seeing (ex ~X MB/min)? Also, is it a CPU or GPU memory leak (I was assuming CPU)?

If I still can't repro it, the GPU model and driver version would help.

**Reply by janmarcheckman** [reply]
Windows 10 and Vulkan is correct
I've been testing with debug build.
Seems to be leaking 0.5 MB/s.
I'm looking at CPU memory.
GPU: NVIDIA GeForce GTX 1070
Driver Version: 31.0.15.4601

**Reply by cwfitzgerald** [reply]
Try with a release build, or with InstanceFlags::empty()

**Reply by janmarcheckman** [reply]
I've graphed the cpu memory usage every minute for 10 minutes for the different combinations of release/debug and InstanceFlags::empty()/default.
![image](https://github.com/user-attachments/assets/5d296316-2cbf-4fdf-bdeb-81d703f17978)

**Reply by teoxoy** [reply]
This sounds like https://github.com/gfx-rs/wgpu/issues/5478. Thanks for investigating!

**Comment by teoxoy** [comment]
I filed https://github.com/gfx-rs/wgpu/issues/6053.
